pkg/pillar/cmd/volumemgr/handlesnapshot.go

```diff
@@ -4,206 +4,286 @@
-	log.Noticef("createVolumeSnapshot: create snapshot for %s", volumeStatus.VolumeID.String())
-	snapshotMeta, timeCreated, err := volumeHandlers.CreateSnapshot()
+func serializeVolumesSnapshotStatus(snapshotID string, status *types.VolumesSnapshotStatus) error {
+	// check if the snapshot directory exists
+	snapshotDir := types.GetSnapshotDir(snapshotID)
+	if _, err := os.Stat(snapshotDir); os.IsNotExist(err) {
+		// Create the directory
+		err = os.MkdirAll(snapshotDir, 0755)
+		if err != nil {
+			log.Errorf("serializeVolumesSnapshotStatus: failed to create snapshot directory %s, %s", snapshotDir, err.Error())
+			return err
+		}
+	}
+	// marshal to JSON
+	statusAsBytes, err := json.Marshal(status)
 	if err != nil {
-		log.Errorf("createVolumeSnapshot: failed to create snapshot for %s, %s", volumeStatus.VolumeID.String(), err.Error())
-		return "", timeCreated, err
+		log.Errorf("serializeVolumesSnapshotStatus: failed to marshal snapshot status for %s, %s", snapshotID, err.Error())
+		return err
 	}
-	log.Noticef("createVolumeSnapshot: successfully created snapshot for %s", volumeStatus.VolumeID.String())
-	return snapshotMeta, timeCreated, nil
+	// Get the filename for the snapshot status
+	volumesSnapshotStatusFile := types.GetVolumesSnapshotStatusFile(snapshotID)
+	// Write the status to the file
+	err = fileutils.WriteRename(volumesSnapshotStatusFile, statusAsBytes)
+	if err != nil {
+		log.Errorf("serializeVolumesSnapshotStatus: failed to write snapshot status for %s, %s", snapshotID, err.Error())
+		return err
+	}
+	return nil
 }
 
-func handleVolumesSnapshotModify(ctxArg interface{}, key string, configArg, _ interface{}) {
-	ctx := ctxArg.(*volumemgrContext)
-	config := configArg.(types.VolumesSnapshotConfig)
-	log.Functionf("handleVolumesSnapshotModify(%s) handles %s", key, config.Action)
+func deserializeVolumesSnapshotStatus(snapshotID string) (*types.VolumesSnapshotStatus, error) {
+	// Get the filename for the snapshot status
+	volumesSnapshotStatusFilename := types.GetVolumesSnapshotStatusFile(snapshotID)
+
+	// check if the volumesSnapshotStatusFile exists
+	if _, err := os.Stat(volumesSnapshotStatusFilename); os.IsNotExist(err) {
+		log.Errorf("deserializeVolumesSnapshotStatus: snapshot status file %s does not exist", volumesSnapshotStatusFilename)
+		return nil, err
+	}
+
+	// open the file
+	volumesSnapshotStatusFile, err := os.Open(volumesSnapshotStatusFilename)
+	if err != nil {
+		log.Errorf("deserializeVolumesSnapshotStatus: failed to open snapshot status file %s, %s", volumesSnapshotStatusFilename, err.Error())
+		return nil, err
+	}
+	defer volumesSnapshotStatusFile.Close()
+
+	// read the raw data
+	data, err := io.ReadAll(volumesSnapshotStatusFile)
+	if err != nil {
+		log.Errorf("deserializeVolumesSnapshotStatus: failed to read snapshot status for %s, %s", snapshotID, err.Error())
+		return nil, err
+	}
+
+	// read to an opaque map to check the fields
```

Comment:
Perhaps this whole part that checks if json files contains required fields and logs warnings about unexpected fields could be generalized into a helper function? I see that you have it twice here and maybe it could be also used for some other current/future persisted jsons.

Reply:
Yes, I thought about it. I can rework it later, especially if you recommend where to put such a function (do we have some generic helpers? `lib/utils`?)
But I would do it later, not blocking the merge of the current PR, if you don't mind.

Reply:
Might make sense to have this error/warning be a separate function in this file - returning an error when you need to bail and otherwise log warnings. That would make the deserialize function itself smaller.